mango/models/pyomo.py

```diff
@@ -0,0 +1,286 @@
+    :param rounding: number of decimal to round.
+    :return: a SuperDict containing the indices and values of the variable.
+    """
+    new_dict = SuperDict(
+        {key: round(safe_value(variable[key]), rounding) for key in variable.keys()}
+    )
+    if clean:
+        return new_dict.clean()
+    else:
+        return new_dict
+
+
+def var_to_table(variable, keys, clean=False, rounding=2):
+    """
+    Generate a table from a pyomo variable.
+
+    :param variable: a pyomo variable.
+    :param keys: the names of the keys of the variable.
+    :param clean: if true, only return non-zero values.
+    :param rounding: number of decimal to round.
+    :return: a table [{Index_1:i1, Index_2:i2, Value:v}...]
+    """
+    var_tup = var_to_sdict(variable, clean=clean, rounding=rounding).to_tuplist()
+    if keys is None:
+        if len(var_tup) > 0 and len(var_tup[0]) > 1:
+            keys = ["Index_" + str(i) for i in range(1, len(var_tup[0]))] + ["Value"]
+        else:
+            keys = ["Value"]
+
+    return Table(var_tup.to_dictlist(keys))
+
+
+def get_status(result):
+    """
+    Return the status of the solution from the result object
+
+    :param result: a pyomo result object
+    :return: the status
+    """
+    return str(result.solver.termination_condition)
+
+
+def get_gap(result):
+    """
+    Return the relative gap of the solution.
+
+    :param result: a pyomo result object
+    :return: the gap
+    """
+    sense = result.Problem.sense
+    lb = result.Problem.lower_bound
+    ub = result.Problem.upper_bound
+
+    if ub == lb:
+        return 0
+    if sense == "minimize":
+        if ub == 0:
+            ub = 10**-6
+        gap = round(100 * (ub - lb) / ub, 3)
+    else:
+        if lb == 0:
+            lb = 10**-6
+        gap = round(100 * (lb - ub) / lb, 3)
+    return abs(gap)
+
+
+def write_cbc_warmstart_file(filename, instance, opt):
+    """
+    This function write a file to be passed to cbc solver as a warmstart file.
+    This function is necessary because of a bug of cbc that does not allow reading warmstart files on windows
+    with absolute path.
+    :param filename: path to the file
+    :param instance: model instance (created with create_instance)
+    :param opt: solver instance (created with solver factory)
+    :return:
+    """
+    opt._presolve(instance)
+    opt._write_soln_file(instance, filename)
+
+
+def variables_to_excel(model_solution, path, clean=True):
+    """
+    Save all the values of the variables in an Excel file.
+
+    :param model_solution: the pyomo solution.
+    :param path: path of the directory where to save the file
+    :param clean: discard variables with values of 0.
+
+    :return: nothing
+    """
+    variables = {
+        k: var_to_table(v, keys=None, clean=clean, rounding=6)
+        for k, v in model_solution.component_map(ctype=pyo.Var).items()
+    }
+    write_excel_light(path + "variables.xlsx", variables)
```

Comment:
Maybe we should add the name of the file as an argument with default name being "variables.xlsx"

Reply:
same for other functions

Reply:
corrected